### PR TITLE
P2209 r0 bulk schedule

### PR DESCRIPTION
**Pull request by LeeHowes**

### CODE REVIEW

File: paper_framework_sources/p2203_bulk_schedule.md [hunk 1]
```diff
@@ -0,0 +1,416 @@
...
+We might implement sort:
+```
+output_sender std::execution::sort(input_sender, executor, policy, comparison_function, range);
+```
+
+where the policy passed to match comparison_function is `par_unseq` but where a complex multi-stage sort needs par internally.
+Here we end up with a question: do we have to restrict the interface to match the implementation? That might be hard.
+The alternative is to transform the executor inside the algorithm.
+Potentially using `require`, or a `with_policy` wrapper.
+The problem is that this is semantically identical to passing the policy with the continuation, but syntactically worse:
```
Comment (lewissbaker) [inline paper_framework_sources/p2203_bulk_schedule.md:323]: I am having trouble following this paragraph. Can you reword or add some extra explainations here?

File: paper_framework_sources/p2203_bulk_schedule.md [hunk 2]
```diff
@@ -0,0 +1,416 @@
...
+ * Composable: rather than building out variants of an algorithm (`post`/`defer`, allocator parameter, different function signatures, per-algorithm shared state, timeouts) composition allows us to build more sophisticated algorthms from simple ones. With inline allocation and laziness we can do this at no cost.
+ * Interoperatable: by defining first the hooks between a `sender` and a `receiver` we define interfaces between different implementors' libraries in the same way that a `range` defines a simple shared concept, irrespective of how an algorithm is implemented. This includes generic support for error handling, value propagation and cancellation.
+ * Cancellable: supporting cancellation in both bulk and non-bulk algorithms is important to a lot of use cases. In [@P0443] the downstream part of cancellation is explicit in the inclusion of `set_done`. The upstream is a query on the receiver.
+
+`bulk_schedule`, compared with `bulk_execute` gives us most of these features.
+This is its power.
+As a trivial example, if we want a blocking operation, we compose bulk_schedule with `sync_wait`:
+```
+sync_wait(bulk_transform(bulk_schedule(ex, 10), [](){...}));
```
[the remark this answers is not in the document]
Comment (LeeHowes) [reply paper_framework_sources/p2203_bulk_schedule.md:41]: You're thinking in terms of libunifex, I was thinking abstractly. Maybe it does make sense to align the two though so libunifex examples make sense to people.

File: paper_framework_sources/p2203_bulk_schedule.md [hunk 3]
```diff
@@ -0,0 +1,416 @@
...
+```
+this requires the `bulk_transform` algorithm, but that is a trivial pass through much like [@P1897]'s `transform`.
+
+If we want to wait with a timeout here, we can add that algorithm:
+```
+sync_wait_for(bulk_transform(bulk_schedule(ex, 10), [](){...}), 10s);
```
[the remark this answers is not in the document]
Comment (LeeHowes) [reply paper_framework_sources/p2203_bulk_schedule.md:47]: Yes, I think the extra algorithm adds confusion. Maybe sync_wait_with_timeout.

File: paper_framework_sources/p2203_bulk_schedule.md [hunk 4]
```diff
@@ -0,0 +1,416 @@
...
+```
+while maintaining a clearly written clue to what we are doing and where the caller will block.
+
+The gap in the [@P2181] definition of `bulk_schedule` is in sequencing of operations.
+This is what we aim to close.
+
+## Proposal TL/DR
+Make `bulk_schedule` symmetric with `schedule` by:
+
+ * Renaming `set_value` to `set_next`, to remove a fundamentally different ordering semantic.
```
[the remark this answers is not in the document]
Comment (LeeHowes) [reply paper_framework_sources/p2203_bulk_schedule.md:57]: Ah yes, you're right that is ambiguous.

File: paper_framework_sources/p2203_bulk_schedule.md [hunk 5]
```diff
@@ -0,0 +1,416 @@
...
+ * Add `set_value` back with the same meaning as in the sender returned from `schedule`.
+ * Define `many_sender` and `many_receiver` to match these definitions.
+ * Propagating the forward progress requirements of `set_value` via a `get_execution_policy` query on the `many_receiver`, consistent with `get_scheduler` as defined in [@P1898].
```
[the remark this answers is not in the document]
Comment (LeeHowes) [reply paper_framework_sources/p2203_bulk_schedule.md]: Yes. As I said earlier we do need a solution to the "is this function safe to call from within this bulk operation" idea but adding this alone doesn't help enough.

File: paper_framework_sources/p2203_bulk_schedule.md [hunk 6]
```diff
@@ -0,0 +1,416 @@
...
+The `many_receiver`’s completion-signal operations have semantic requirements that are collectively known as the `many_receiver` contract, described below:
+
+None of a `many_receiver`’s completion-signal operations shall be invoked before `execution::start` has been called on the operation state object that was returned by `execution::connect` to connect that `many_receiver` to a `many_sender`.
+
+Once `execution::start` has been called on the operation state object, `set_next` shall be called for each index in some iteration space and parameterised with that index, under the restriction of the policy returned by a call to `get_execution_policy` on the `many_receiver`.
+
+All of the `many_receiver`'s `set_next` calls shall happen-before any of the `many_receiver`'s completion-signal operations.
+Exactly one of the receiver’s completion-signal operations shall complete non-exceptionally before the receiver is destroyed.
+
+If any call to `execution::set_next` or `execution::set_value` exits with an exception, it is still valid to call `execution::set_error` or `execution::set_done` on the `many_receiver`.
```
[the remark this answers is not in the document]
Comment (LeeHowes) [reply paper_framework_sources/p2203_bulk_schedule.md:274]: I would leave it up to the sender.  That's why I kept this constraint simple.

Now, what the bulk_scheduler should do is another question - but if we over constrain it it'll make some people unhappy. Might be better to allow *any* exception to propagate out as "there was an error" with no direct link.

File: paper_framework_sources/p2203_bulk_schedule.md [hunk 7]
```diff
@@ -354,14 +352,21 @@ based on prior work, each of these user-visible algorithms would take a policy.
 They are all user-facing as well as implementation-details.
 In the `get_execution_policy` model each of these can communicate a policy to the prior algorithm.
 
-These algorithms are communicating single values rather than bulk.
-If we want to be sure that we can implement a completion signal safely, that is useful because it allows the executor to guarantee safe communication with the next executor.
+We do realise that execution policy is a broader concept than just forward progress - it may be that we need to separate concerns here.
+The aspect of the execution policy that is a minimum requirement for the code being passed in - the way they are used in the current set of defined policies on the parallel algorithms - may be abstracted in this way while the wider policy is separate and describing execution may be useful.
+In that case `get_forward_progress_requirement` might be a better name.
```
Comment (lewissbaker) [inline paper_framework_sources/p2203_bulk_schedule.md:357]: I think there is possibly also another aspect bound up in the constraints/policy in addition to forward-progress, and that is "thread safety". e.g. if a receiver returns a policy of `par`, then it's not only saying that it requires concurrent forward progress guarantee for threads that call it, but also that it is safe to call concurrently from multiple threads.

File: paper_framework_sources/p2203_bulk_schedule.md [hunk 8]
```diff
@@ -0,0 +1,416 @@
...
+      concept many_sender =
+        move_constructible<remove_cvref_t<S>> &&
+        !requires {
+          typename sender_traits<remove_cvref_t<S>>::__unspecialized; // exposition only
+        };
+
+    template<class S, class R>
+      concept many_sender_to =
+        many_sender<S> &&
+        many_receiver<R> &&
```
[the remark this answers is not in the document]
Comment (LeeHowes) [reply paper_framework_sources/p2203_bulk_schedule.md:293]: Yes. The problem is that the receiver concept is very simple. I really need a many_receiver_of  but then i need the value type which many_sender_to doesn't directly provide.

File: paper_framework_sources/p2203_bulk_schedule.md [hunk 9]
```diff
@@ -0,0 +1,416 @@
...
+        requires (S&& s, R&& r) {
+          execution::connect((S&&) s, (R&&) r);
+        };
+```
+
+None of these operations shall introduce data races as a result of concurrent invocations of those functions from different threads.
+
+A `many_sender` type’s destructor shall not block pending completion of the submitted function objects. [Note: The ability to wait for completion of submitted function objects may be provided by the associated execution context. –end note]
```
[the remark this answers is not in the document]
Comment (LeeHowes) [reply paper_framework_sources/p2203_bulk_schedule.md:301]: That's a bug in the sender wording in P0443.

File: paper_framework_sources/p2203_bulk_schedule.md [hunk 10]
```diff
@@ -0,0 +1,416 @@
...
+If any call to `execution::set_next` or `execution::set_value` exits with an exception, it is still valid to call `execution::set_error` or `execution::set_done` on the `many_receiver`.
+If all calls to `execution::set_next` complete successfully, it is valid to call `execution::set_value` on the `many_receiver`.
+
+Once one of a `many_receiver`’s completion-signal operations has completed non-exceptionally, the `many_receiver` contract has been satisfied.
+
+
+## Concepts many_sender and many_sender_to
+XXX TODO The many_sender and many_sender_to concepts…
+```
+    template<class S>
```
[the remark this answers is not in the document]
Comment (LeeHowes) [reply paper_framework_sources/p2203_bulk_schedule.md:283]: Yes... probably, but the whole sender concepts section is XXX TODO in P0443 anyway. I need Eric's expertise to get this right.

File: paper_framework_sources/p2203_bulk_schedule.md [hunk 11]
```diff
@@ -353,7 +395,7 @@ They are all user-facing as well as implementation-details.
 In the `get_execution_policy` model each of these can communicate a policy to the prior algorithm.
 
 We do realise that execution policy is a broader concept than just forward progress - it may be that we need to separate concerns here.
-The aspect of the execution policy that is a minimum requirement for the code being passed in - the way they are used in the current set of defined policies on the parallel algorithms - may be abstracted in this way while the wider policy is separate and describing execution may be useful.
+The aspect of the execution policy that is a statement of what the most relaxed conditions in which the function is safe to callthe way they are used in the current set of defined policies on the parallel algorithms, may be abstracted in this way while the wider policy is separate and describing execution may be useful.
```
Comment (lewissbaker) [inline paper_framework_sources/p2203_bulk_schedule.md:398]: `callthe` -> `call the`

**Pull request merged**